File: 0205-isomorphic-strings/0205-isomorphic-strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if s is None or len(s) == 0 or t is None or len(t)==0:
            return False
        
        sMap = dict()
        tMap = dict()
        
        for i in range(0, len(s)):
            sChar = s[i]
            tChar = t[i]
            
            if sChar in sMap.keys():
                # print(sChar, sMap)
                if sMap.get(sChar) != t[i]:
                    return False
            else:
                sMap[sChar] = tChar
            
            if tChar in tMap.keys():
                # print(tChar, tMap)
                if tMap.get(tChar) != s[i]:
                    return False
            else:
                tMap[tChar] = sChar
        
        return True
    
    
                    

File: 0205-isomorphic-strings/0205-isomorphic-strings.java
class Solution {
    public boolean isIsomorphic(String s, String t) {
        if(s == null || s.length() == 0 || t == null || t.length() == 0){
            return true;
        }

        if(s.length() != t.length()){
            return false;
        }

        Map<Character, Character> sMap = new HashMap<>();
        Map<Character, Character> tMap = new HashMap<>();

        for(int i = 0; i < s.length(); i++){
            char sChar = s.charAt(i);
            char tChar = t.charAt(i);

            if(!sMap.containsKey(sChar)){
                sMap.put(sChar, tChar);
            }
            else{
                if(sMap.get(sChar) != tChar){
                    return false;
                }
            }

            if(!tMap.containsKey(tChar)){
                tMap.put(tChar, sChar);
            }
            else{
                if(tMap.get(tChar) != sChar){
                    return false;
                }
            }
        }

        return true;

    }
}

// #Time Complexity: O(n)
// #Space Complexity: O(n)